File: core/services/performance/memory_monitor/memory_record.cc
// Copyright 2025 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.

#include "core/services/performance/memory_monitor/memory_record.h"

namespace lynx {
namespace tasm {
namespace performance {

MemoryRecord::MemoryRecord(MemoryCategory category, float size_kb)
    : category_(std::move(category)), size_kb_(size_kb) {}

MemoryRecord::MemoryRecord(
    MemoryCategory category, float size_kb,
    std::unique_ptr<std::unordered_map<std::string, std::string>> detail)
    : category_(std::move(category)),
      size_kb_(size_kb),
      detail_(std::move(detail)) {}

MemoryRecord::MemoryRecord(
    MemoryCategory category, float size_kb, int32_t instance_count,
    std::unique_ptr<std::unordered_map<std::string, std::string>> detail)
    : category_(std::move(category)),
      size_kb_(size_kb),
      instance_count_(instance_count),
      detail_(std::move(detail)) {}

MemoryRecord& MemoryRecord::operator+=(const MemoryRecord& other) {
  size_kb_ += other.size_kb_;
  instance_count_ += other.instance_count_;
  // Merge detail_
  if (!other.detail_) {
    return *this;
  }
  if (!detail_) {
    // If we have no detail, create a copy of other's
    detail_ = std::make_unique<std::unordered_map<std::string, std::string>>(
        *other.detail_);
    return *this;
  }
  for (const auto& [key, value] : *other.detail_) {
    (*detail_)[key] = value;
  }
  return *this;
}

MemoryRecord& MemoryRecord::operator-=(const MemoryRecord& other) {
  size_kb_ -= other.size_kb_;
  instance_count_ -= other.instance_count_;
  // Deduplicate detail_
  if (!other.detail_ || !detail_) {
    return *this;
  }
  for (const auto& [key, value] : *other.detail_) {
    detail_->erase(key);
  }
  return *this;
}

}  // namespace performance
}  // namespace tasm
}  // namespace lynx

File: platform/android/lynx_android/src/main/java/com/lynx/tasm/performance/memory/MemoryRecord.java
// Copyright 2025 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.

package com.lynx.tasm.performance.memory;

import java.util.Map;

/**
 * @brief Represents a memory record with a category, size in kilobytes, and detail information.
 */
public class MemoryRecord {
  /** The category of the memory record. like "image", "vm", "element" etc. */
  private String mCategory;
  /** The size of the memory record in kilobytes. */
  private float mSizeKb;
  /** The detail information of the memory record. */
  private Map<String, String> mDetail = null;

  public MemoryRecord(String category, float sizeKb, Map<String, String> detail) {
    mCategory = category;
    mSizeKb = sizeKb;
    mDetail = detail;
  }

  public String getCategory() {
    return mCategory;
  }

  public float getSizeKb() {
    return mSizeKb;
  }

  public Map<String, String> getDetails() {
    return mDetail;
  }
}
